src/modules/navigator/mission.cpp

```diff
@@ -1071,21 +1120,36 @@ Mission::set_mission_items()
 		set_current_mission_item();
 	}
 
-	if (_mission_item.autocontinue && get_time_inside(_mission_item) < FLT_EPSILON) {
-		/* try to process next mission item */
-		if (has_next_position_item) {
+	// If the mission item under evaluation contains a gate,
+	// then we need to check if we have a next position item so
+	// the controller can fly the correct line between the
+	// current and next setpoint
+	if (item_contains_gate(_mission_item)) {
+		if (has_after_next_position_item) {
```

Comment:
I'm trying to understand why this logic is needed twice :thinking:.

Reply:
I didn't want to refactor it in parallel, but basically one area is concerned with current waypoints and the one with next. This whole function needs to be broken down accordingly to make it readable again.

Reply:
Right, sadly that won't be done anytime soon.